examples/vision/xray_classification_with_tpus.py

```diff
@@ -0,0 +1,489 @@
+many images we have in our validation set. Verify that the ratio of images is 80:20.
+"""
+
+TRAIN_IMG_COUNT = tf.data.experimental.cardinality(train_list_ds).numpy()
+print("Training images count: " + str(TRAIN_IMG_COUNT))
+
+VAL_IMG_COUNT = tf.data.experimental.cardinality(val_list_ds).numpy()
+print("Validating images count: " + str(VAL_IMG_COUNT))
+
+"""
+As expected, we have two labels for our images.
+"""
+
+CLASS_NAMES = np.array(
+    [
+        str(tf.strings.split(item, os.path.sep)[-1].numpy())[2:-1]
+        for item in tf.io.gfile.glob(str(GCS_PATH + "/chest_xray/train/*"))
+    ]
+)
+CLASS_NAMES
+
+"""
+Currently our dataset is just a list of filenames. We want to map each filename to the
+corresponding (image, label) pair. The following methods will help us do that.
+
+As we only have two labels, we will rewrite the label so that `1` or `True` indicates
+pneumonia and `0` or `False` indicates normal.
+"""
+
+
+def get_label(file_path):
+    # convert the path to a list of path components
+    parts = tf.strings.split(file_path, os.path.sep)
+    # The second to last is the class-directory
+    return parts[-2] == "PNEUMONIA"
+
+
+"""
+The images originally have values that range from [0, 255]. CNNs work better with smaller
+numbers so we will scale this down.
+"""
+
+
+def decode_img(img):
+    # convert the compressed string to a 3D uint8 tensor
+    img = tf.image.decode_jpeg(img, channels=3)
+    # Use `convert_image_dtype` to convert to floats in the [0,1] range.
+    img = tf.image.convert_image_dtype(img, tf.float32)
+    # resize the image to the desired size.
+    return tf.image.resize(img, IMAGE_SIZE)
+
+
+def process_path(file_path):
+    label = get_label(file_path)
+    # load the raw data from the file as a string
+    img = tf.io.read_file(file_path)
+    img = decode_img(img)
+    return img, label
+
+
+train_ds = train_list_ds.map(process_path, num_parallel_calls=AUTOTUNE)
+
+val_ds = val_list_ds.map(process_path, num_parallel_calls=AUTOTUNE)
+
+"""
+Let's visualize the shape of an (image, label) pair.
+"""
+
+for image, label in train_ds.take(1):
+    print("Image shape: ", image.numpy().shape)
+    print("Label: ", label.numpy())
+
+"""
+Load and format the test data as well.
+"""
+
+test_list_ds = tf.data.Dataset.list_files(str(GCS_PATH + "/chest_xray/test/*/*"))
+TEST_IMAGE_COUNT = tf.data.experimental.cardinality(test_list_ds).numpy()
+test_ds = test_list_ds.map(process_path, num_parallel_calls=AUTOTUNE)
+test_ds = test_ds.batch(BATCH_SIZE)
+
+TEST_IMAGE_COUNT
+
+"""
+# 3. Visualize the dataset
+
+First, let's use buffered prefetching so we can yield data from disk without having I/O
+become blocking.
+"""
+
+
+def prepare_for_training(ds, cache=True, shuffle_buffer_size=1000):
+    # This is a small dataset, only load it once, and keep it in memory.
+    # use `.cache(filename)` to cache preprocessing work for datasets that don't
+    # fit in memory.
+    if cache:
```

Comment:
In general for an image dataset it is not recommend to cache it in memory. Prefer downloading the files to local disk and reading them from disk without caching.

Reply:
For TPUs, I don't think we can access local files. All files have to be access through a GCS path. This might be why `image_dataset_from_directory` does not work with TPUs.

Reply:
Ok, that makes sense. Please include an explanation about how large image datasets should not be cached in memories, but here you do it because you want to train on TPU and the dataset is not very large.